### PR TITLE
Tighten source-spans to token end position

### PR DESCRIPTION
**Pull request by nwolverson**
An attempt at #1702. Don't seem to get much help from parsec here, this is bit of a fudge - and (at least?) the Parsec parsers used for string & number literals still consume whitespace (`lexeme` itself eats whitespace). I don't see any way around that other than not using `LanguageDef`.

### CODE REVIEW

**Comment by hdgarrood** [comment]
How easy is it to add tests for this?

File: src/Language/PureScript/Parser/Lexer.hs [hunk 1]
```diff
@@ -153,6 +153,8 @@ prettyPrintToken (HoleLit name)    = "?" ++ name
 
 data PositionedToken = PositionedToken
   { ptSourcePos :: P.SourcePos
+  , ptEndPos :: P.SourcePos
+  , ptPrevEndPos :: Maybe P.SourcePos
```
Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: Could you please add some Haddock comments to these fields?

File: src/Language/PureScript/Parser/Declarations.hs [hunk 2]
```diff
@@ -281,6 +285,13 @@ parseModulesFromFiles toFilePath input = do
   inParallel :: [Either P.ParseError (k, [Module])] -> [Either P.ParseError (k, [Module])]
   inParallel = withStrategy (parList rseq)
 
+  updatePositions :: [PositionedToken] -> [PositionedToken]
```
Comment (paf31) [inline src/Language/PureScript/Parser/Declarations.hs]: Why not put this function in the lexer module, or even fuse it into the lexer function?

**Comment by nwolverson** [comment]
On tests, I think specific parser tests could be done invoking `parseModules` and checking the results, maybe without external files. Not sure what's appropriate here, that's a bit different from existing tests? Input required.

**Comment by paf31** [comment]
👍 LGTM, thanks!

**Comment by nwolverson** [comment]
Squashed and rebased to 0.9. Would you like me to open a new PR against 0.9 or will you merge manually?

**Comment by paf31** [comment]
Yes please.

**Comment by nwolverson** [comment]
https://github.com/purescript/purescript/pull/2077

**Pull request closed without merging**